### PR TITLE
IGNITE-20303 Fixed "Raft group on the node is already started"

### PR DESCRIPTION
**Pull request by sergeyuttsel**
https://issues.apache.org/jira/browse/IGNITE-20303

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 1]
```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
...
+        byte[] bytesPendingAssignments = ByteUtils.toBytes(pendingAssignments);
+        byte[] bytesPlannedAssignments = ByteUtils.toBytes(plannedAssignments);
+
+        TablePartitionId partId = new TablePartitionId(1, 0);
+
+        ByteArray partAssignmentsStableKey = stablePartAssignmentsKey(partId);
+        ByteArray partAssignmentsPendingKey = pendingPartAssignmentsKey(partId);
+        ByteArray partAssignmentsPlannedKey = plannedPartAssignmentsKey(partId);
+
+        SimpleCondition condition = exists(partAssignmentsStableKey);
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:620]: I think you can use simple `MetaStorageManager#putAll`

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 2]
```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
...
+
+        updateAssignments();
+
+        checkAssignments();
+
+        verifyThatRaftNodesAndReplicasWereStartedOnlyOnce();
+    }
+
+    private void clearSpyInvocations() {
+        clearInvocations(getNode(0).getRaftManager());
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:582]:
```
IntStream.range(0, NODE_COUNT).forEach(i -> {
            clearInvocations(getNode(i).getRaftManager());
            clearInvocations(getNode(i).getReplicaManager());
         });
```
?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 3]
```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
...
+                ops().yield(false)
+        );
+
+        MetaStorageManager metaStorageMgr = getNode(0).getMetaStorageManager();
+
+        assertTrue(metaStorageMgr.invoke(iif).get(AWAIT_TIMEOUT_MILLIS, MILLISECONDS).getAsBoolean());
+    }
+
+    private void verifyThatRaftNodesAndReplicasWereStartedOnlyOnce() throws Exception {
+        verify(getNode(0).getRaftManager(), times(1))
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:639]:
```
 for (int i = 0; i < NODE_COUNT; i++) {
    verify(getNode(i).getRaftManager(), times(1))
            .startRaftGroupNode(any(), any(), any(), any(), any(RaftGroupOptions.class));
    verify(getNode(i).getReplicaManager(), times(1)).startReplica(any(), any(), any(), any(), any());
}
```
?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 4]
```diff
@@ -962,6 +1123,22 @@ TablePartitionId getTablePartitionId(String tableName, int partitionId) {
 
             return new TablePartitionId(internalTable.tableId(), partitionId);
         }
+
+        public String getName() {
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:1127]: Node is a private class only for this test class. I would prefer to just make the fields public, intead of spam test itself with these getters

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 5]
```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
...
+
+        createTable(node, ZONE_NAME, TABLE_NAME);
+
+        assertTrue(waitForCondition(() -> getPartitionClusterNodes(node, 0).size() == 1, AWAIT_TIMEOUT_MILLIS));
+
+        alterZone(node, ZONE_NAME, 3);
+
+        checkAssignments();
+
+        updateAssignments();
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:574]: I would prefer to more specific naming, like `directUpdateMetastoreRebalanceAssignmentKeys` to show the real intention of this update.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 6]
```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
...
+
+        checkAssignments();
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:576]: Again, I would prefer `checkPartitionAssignmentsSyncedAndRebalanceKeysEmpty`

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 1 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Fixed.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 2 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Yes, thanks)

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 3 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Yes, thanks)

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 4 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Fixed.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 5 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Fixed.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 6 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java]: Fixed.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java [hunk 7]
```diff
@@ -535,13 +549,102 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
         checkInvokeDestroyedPartitionStorages(newNode, TABLE_NAME, 0);
     }
 
+    @Test
+    void testRebalanceWithTheSameNodes() throws Exception {
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:553]: @sergeyuttsel could you please add the comment about this case? Not for formal reasons, but I'm afraid that this case is a really hard for understanding. You can copy part of my description from our discussions

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java (hunk 7 above)
Comment (sergeyuttsel) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java:571]: Fixed.

**Review by kgusakov — APPROVED** [review]

**Pull request merged**